tmt/utils.py

```diff
@@ -3859,6 +3859,76 @@ def dataclass_normalize_field(
     return value
 
 
+# TODO: wouldn't it be nice if typing would support some kind of `typeof(...)`
+# helper? it doesn't, yet. https://github.com/python/typing/issues/769
+class OptionGetter(Protocol):
+    def __call__(self, option: str, default: Optional[Any] = None) -> Any:
+        pass
+
+
+def dataclass_normalize_options(
+        *,
+        container: Any,
+        option_getter: OptionGetter,
+        keys: Optional[List[str]] = None,
+        preserve_modified: bool = True,
+        logger: tmt.log.Logger,
+        ) -> None:
+    """
+    Normalize and assign values from CLI options to container fields.
+
+    :param container: container to update.
+    :param option_getter: a callable to provide a value for a given option.
+        It is called once for each option name, with an optional default
+        to return if the option was not specified.
+    :param keys: optional list of keys to update. If not set, all keys
+        are considered.
+    :param preserve_modified: if set, only options whose value is not equal
+        to field's default value would be saved in the container.
+    :param logger: used for logging.
+    """
+
+    keys = keys or list(container.keys())
+
+    for keyname in keys:
+        # Checks below try to prevent default options of CLI options from
+        # overwriting values already consumed from fmf nodes. This is no
+        # simple task, since options have defaults, they are always present
+        # in the set of options provided by Click, therefore we have no way
+        # how to check whether the option was specified or not. Instead, we
+        # are "guessing", and ignoring default-ish values.
+
+        # Some keys are not even changeable via command line. Filter those
+        # out first. With a special default value, we can detect options
+        # that were never seen by Click.
+        value = option_getter(tmt.utils.key_to_option(keyname), default=dataclasses.MISSING)
+
+        if value is dataclasses.MISSING:
+            continue
+
+        if preserve_modified:
+            # If the value we get for an option is the default value, good,
+            # no need to bother - as such, it's either already stored in
+            # `self.data`, or it has been replaced with whatever came from an
+            # fmf node.
+            if value == container._default(keyname):
+                continue
```

Comment:
All of this is just wrong... It's all guessing, and it's a big smoking pile of corner cases. Let me find out what exactly breaks when this particular check is dropped, I bet there's something just waiting to bite us :)

Reply:
Summary from the call: We need to find a reliable way how to detect if a flag option was provided on the command line or not (versus set as the default). Moving to the next release because of the regression (unable to override non-default config value with default value provided on the command line).

Reply:
Right, it's the boolean value of these defaults. TL;DR, there's no way for code to tell whether `-i` was used on the command line, or whether the `fedora` string comes from the default value. For the code, these two look the same: `-h container` vs `-h container -i fedora`, both would present `fedora` as the value of the `--image` CLI option. Now, in the first case, we do not want to overwrite fmf source, in the second we do want to. How to tell the difference?

As briefly discussed, there is a couple of options, none trivial or immediately available. I'd try to look into introducing a callback that would notify us when an option was actually encountered on a command line, and we'd then simply use the value no matter what it ould be. Postponing this PR to avoid bigger harm.